cmd/compose/top.go

```diff
@@ -63,30 +66,60 @@ func runTop(ctx context.Context, dockerCli command.Cli, backend api.Service, opt
 		return containers[i].Name < containers[j].Name
 	})
 
+	header, entries := collectTop(containers)
+	return topPrint(dockerCli.Out(), header, entries)
+}
+
+func collectTop(containers []api.ContainerProcSummary) (topHeader, []topEntries) {
+	// map column name to its header (should keep working if backend.Top returns
+	// varying columns for different containers)
+	header := topHeader{"SERVICE": 0}
+
+	// assume one process per container and grow if needed
+	entries := make([]topEntries, 0, len(containers))
+
 	for _, container := range containers {
-		_, _ = fmt.Fprintf(dockerCli.Out(), "%s\n", container.Name)
-		err := psPrinter(dockerCli.Out(), func(w io.Writer) {
-			for _, proc := range container.Processes {
-				info := []interface{}{}
-				for _, p := range proc {
-					info = append(info, p)
-				}
-				_, _ = fmt.Fprintf(w, strings.Repeat("%s\t", len(info))+"\n", info...)
+		for _, proc := range container.Processes {
+			entry := topEntries{"SERVICE": container.Name}
```

Comment:
container name is not service name - you should use container label `com.docker.compose.service` to retrieve service name (and make display shorter)

Reply:
But then I should probably also include `com.docker.compose.container-number`, no?

Should I recombine them (service + "-" + container-number), or should this be a separate column (titled "#" to keep it short)?

Reply:
a separate column for replica # makes sense

Reply:
Hm. Apart from `(api.Service).Ps()`, no other method returns the labels.

AFAICT, I have these options:

- Just call `Ps()` and discard all but the labels.
- Extend `api.ContainerProcSummary` with `ServiceName` and `ReplicaID` fields (and adjust `(*composeService).Top()` to fill these).
- Extend `api.ContainerProcSummary` with a `Labels` field (and adjust `(*composeService).Top()` to fill it).
- Keep `container.Name` as-is.

Any preference?

Reply:
I'm tending to prefer option (2) or (3), as `(*composeService).Top()` already has the labels available.

Option (1) would cause additional API calls (looks like one for each container, not sure).

Reply:
There's also a `getContainerNameWithoutProject()` helper function, which is used in various other places.

Maybe this is a better choice (in terms of consistency) for `(ContainerProcSummary).Name`, as it defaults to `getCanonicalContainerName()` (the current value for `(ContainerProcSummary).Name`)?

Reply:
extend `ContainerProcSummary` to match your needs. `getContainerNameWithoutProject` is not 100% safe as service may declare a `container_name` that can confuse this function.

Reply:
I'm now passing the container labels unchanged to the commandlet.

- It doesn't make much sense filtering the labels in `(*composeService).Top()` (any other consumer would wonder where the other labels are).
- Promoting the service and container-number labels to proper fields on `ContainerProcSummary` also feels arbitrary. Also, the container-number ought to be of type `int`, which implies `strconv` both from and to string...